packages/babel-plugin-transform-parameters/src/rest.js

```diff
@@ -153,6 +153,17 @@ const memberExpressionOptimisationVisitor = {
     }
   },
 };
+
+function getParamsCount(node) {
+  let count = node.params.length;
+  // skip the first parameter if it is a TypeScript 'this parameter'
+  const p0 = node.params[0];
+  if (p0 && t.isIdentifier(p0) && p0.name === "this") {
```

Comment:
If `count` is guarded, we don't have to check `p0` because each node of `params` must not be nullish.

Also you can simplify the `isIdentifier` check to
```js
t.isIdentifier(p0, { name: "this" })
```